pkg/controller/garbagecollector/codec.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type compressingCodec struct {
+	dynamic.Codec
+}
+
+func NewCompressingCodec() compressingCodec {
+	return compressingCodec{dynamic.Codec{}}
+}
+
+// Decode calls the embedded Codec to decode data. If the decoded object is
+// runtime.Unstructured or runtime.UnstructuredList, Decode only keep the
+// APIVersion, Kind, and ObjectMeta fields.
+func (c compressingCodec) Decode(data []byte, gvk *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
+	ret, gvk, err := c.Codec.Decode(data, gvk, into)
```

Comment:
You mean just decode to something like

```go
type MetaOnly struct {  //TODO pick better name
	unversioned.TypeMeta `json:",inline"`
	v1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`  // is this standard enough to be a common type?
}
```
?

It would be hard to have the json package to decode to `Unstructured` but skip fields. But the json decoding package let's you easily do  this with structs.

The only problem then is that the dynamic client explicitly decodes into `runtime.Unstructured` or `runtime.List`. It might be a bit much, but possibly we could fix this by defining a dynamic codec interfaces that can create objects and lists for us to decode into. It would also allow us to know if a true dynamic codec was already specified in the restclient config, and we wouldn't need to override it. Something like:

```go
type DynamicCode interface {
	runtime.Codec
	NewDynamicObject() runtime.Object
	NewDynamicList() runtime.Object
}
```

The current dynamic codec could return `runtime.Unstructured` and `runtime.UnstructuredList` respectively. And this new meta-only codec could return structs that have only the meta fields.

This is probably way out of scope though.